### PR TITLE
Update upper bound for numpy.

### PR DESCRIPTION
**Pull request by tvalentyn**
Due to https://github.com/numpy/numpy/pull/22004, updating one of the unit tests that creates an nd.array with lists of variable length.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/python/setup.py [hunk 1]
```diff
@@ -246,7 +246,7 @@ def get_portability_package_data():
         'grpcio>=1.33.1,!=1.48.0,<2',
         'hdfs>=2.1.0,<3.0.0',
         'httplib2>=0.8,<0.21.0',
-        'numpy>=1.14.3,<1.23.0',
+        'numpy>=1.14.3,<1.25.0',
```
Comment (tvalentyn) [inline sdks/python/setup.py:249]:
```suggestion
        'numpy>=1.24.0,<1.25.0',
```

**Comment by tvalentyn** [comment]
Run Python 3.8 PostCommit

**Comment by tvalentyn** [comment]
Another issue: 
```
10:41:08 Traceback (most recent call last):
10:41:08   File "apache_beam/runners/common.py", line 1465, in apache_beam.runners.common.DoFnRunner._invoke_lifecycle_method
10:41:08   File "apache_beam/runners/common.py", line 551, in apache_beam.runners.common.DoFnInvoker.invoke_setup
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/ml/inference/base.py", line 427, in setup
10:41:08     self._model = self._load_model()
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/ml/inference/base.py", line 420, in _load_model
10:41:08     return self._shared_model_handle.acquire(load)
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/utils/shared.py", line 305, in acquire
10:41:08     return _shared_map.acquire(self._key, constructor_fn, tag)
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/utils/shared.py", line 246, in acquire
10:41:08     result = control_block.acquire(constructor_fn, tag)
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/utils/shared.py", line 139, in acquire
10:41:08     result = constructor_fn()
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/ml/inference/base.py", line 409, in load
10:41:08     model = self._model_handler.load_model()
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/ml/inference/base.py", line 167, in load_model
10:41:08     return self._unkeyed.load_model()
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/ml/inference/tensorrt_inference.py", line 267, in load_model
10:41:08     return TensorRTEngine(engine)
10:41:08   File "/usr/local/lib/python3.8/dist-packages/apache_beam/ml/inference/tensorrt_inference.py", line 134, in __init__
10:41:08     'dtype': np.dtype(trt.nptype(dtype)),
10:41:08   File "/usr/local/lib/python3.8/dist-packages/tensorrt/__init__.py", line 165, in nptype
10:41:08     bool: np.bool,
10:41:08   File "/usr/local/lib/python3.8/dist-packages/numpy/__init__.py", line 284, in __getattr__
10:41:08     raise AttributeError("module {!r} has no attribute "
10:41:08 AttributeError: module 'numpy' has no attribute 'bool'
10:41:08 
```

**Comment by tvalentyn** [comment]
Rootcause tracked in: https://github.com/NVIDIA/TensorRT/issues/2557

**Comment by jrmccluskey** [comment]
Error message on the precommits: 

```
Container does not include required Beam dependencies or has conflicting dependencies. If Beam dependencies have changed, you need to regenerate base_image_requirements.txt files. See: https://s.apache.org/beam-python-requirements-generate
The command '/bin/sh -c pip check || (echo "Container does not include required Beam dependencies or has conflicting dependencies. If Beam dependencies have changed, you need to regenerate base_image_requirements.txt files. See: https://s.apache.org/beam-python-requirements-generate" && exit 1)' returned a non-zero code: 1
```

**Comment by tvalentyn** [comment]
thanks. that's WAI. Bot sent it for review too early. I have intentionally temporarily increased the lower bound temporarily to have a meaningful postcommit signal in https://ci-beam.apache.org/job/beam_PostCommit_Python38_PR/674/ (still running)

**Comment by tvalentyn** [comment]
Run Python 3.8 PostCommit

**Comment by tvalentyn** [comment]
Run Python 3.8 PostCommit

**Comment by damccorm** [comment]
stop reviewer notifications

**Comment by damccorm** [comment]
@tvalentyn I'm happy to review, but I'm stopping the bot since its not ready yet

**Comment by tvalentyn** [comment]
cc: @jli

**Comment by tvalentyn** [comment]
Interactive runner tests need a fix in facets-overview: https://github.com/PAIR-code/facets/pull/250

**Comment by tvalentyn** [comment]
Run Python 3.8 PostCommit

**Comment by tvalentyn** [comment]
Dataframes (ParquetIO) test requires numpy<1.20.0, this is a preexisting issue: https://github.com/apache/beam/blob/428ec97e30cc6587c2ae0f81d3ba44a8a9b34f93/sdks/python/tox.ini#L286 , and shouldn't fail once I relax the lower bound.

**Comment by tvalentyn** [comment]
> Interactive runner tests need a release from facets-overview: https://github.com/PAIR-code/facets/pull/250

This has been done

**Comment by tvalentyn** [comment]
>> Interactive runner tests need a release from facets-overview: https://github.com/PAIR-code/facets/pull/250

>This has been done

needs one more patch... https://github.com/PAIR-code/facets/pull/251

**Comment by tvalentyn** [comment]
> needs one more patch... PAIR-code/facets#251

should be released now.

One more issue. Looks like some DF tests fail with pandas 1.11. The bug seems to be in Pandas 1.11, the change in behavior is caused by following line of code:

At some point the test is executing something akin to:

```
import numpy as np
x = np.empty(0, dtype='int64')
x.fill(np.NaN)
```

which now fails. I am considering bumping up the lower bound of Pandas, the affected version is 2+ yrs old.

**Comment by tvalentyn** [comment]
Pandas 1.1.x issue is reproducible in: `pd.DataFrame(columns=pd.CategoricalIndex(['A'], dtype='category'), dtype='int64')` , so bumping pandas is appropriate.

**Comment by tvalentyn** [comment]
> Pandas 1.1.x issue is reproducible in: `import pandas as pd; pd.DataFrame(columns=pd.CategoricalIndex(['A'], dtype='category'), dtype='int64')` , so bumping pandas is appropriate.

Looking closer, it seems that minimal pandas version that doesn't fail on recent version of numpy is pandas==1.4.3

**Comment by tvalentyn** [comment]
note that pandas==1.4.3 is not available on Python 3.7, but neither is numpy==1.24.x.

**Comment by tvalentyn** [comment]
Also bumping up lower bounds for pyarrow, for similar reason.

**Comment by tvalentyn** [comment]
Ok, this should now be ready for review, @damccorm

File: sdks/python/test-suites/tox/py38/build.gradle [hunk 2]
```diff
@@ -85,21 +73,9 @@ test.dependsOn "testPy38pyarrow-9"
...
-preCommitPyCoverage.dependsOn "testPy38pandas-12"
-
-toxTask "testPy38pandas-13", "py38-pandas-13", "${posargs}"
-test.dependsOn "testPy38pandas-13"
-preCommitPyCoverage.dependsOn "testPy38pandas-13"
-
 toxTask "testPy38pandas-14", "py38-pandas-14", "${posargs}"
 test.dependsOn "testPy38pandas-14"
-preCommitPyCoverage.dependsOn "testPy38pandas-14"
+preCommitPy38.dependsOn "testPy38pandas-14"
```
Comment (tvalentyn) [inline sdks/python/test-suites/tox/py38/build.gradle:78]:
```suggestion
preCommitPyCoverage.dependsOn "testPy38pandas-14"
```

File: sdks/python/apache_beam/ml/inference/tensorrt_inference.py [hunk 3]
```diff
@@ -122,7 +122,16 @@ def __init__(self, engine: trt.ICudaEngine):
     self.outputs = []
     self.gpu_allocations = []
     self.cpu_allocations = []
-    """Setup I/O bindings."""
+
+    # TODO(https://github.com/NVIDIA/TensorRT/issues/2557):
```
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/tensorrt_inference.py:126]: Should we have an issue on our end to track this?

**Review by damccorm — APPROVED** [review]

File: sdks/python/setup.py [hunk 4]
```diff
@@ -343,15 +345,12 @@ def get_portability_package_data():
...
-        # This can be removed once dill is updated to version > 0.3.5.1
-        # Issue: https://github.com/apache/beam/issues/23566
+        # Exclude pandas<=1.4.2 since it doesn't work with numpy 1.24.x.
         # Exclude 1.5.0 and 1.5.1 because of
         # https://github.com/pandas-dev/pandas/issues/45725
           'dataframe': [
-            'pandas>=1.0,<1.6,!=1.5.0,!=1.5.1;python_version<"3.10"',
-            'pandas>=1.4.3,<1.6,!=1.5.0,!=1.5.1;python_version>="3.10"'
+            'pandas<1.4.0;python_version=="3.7"',
+            'pandas>=1.4.3,!=1.5.0,!=1.5.1,<1.6;python_version>="3.8"',
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/setup.py:353]: True. The last 1.3.x version was released Dec 2021 though. The patch versions of 1.4.x are probably interchangeable to users.

Also customers could try using an older version of Beam.  We can make a change to release notes it if would help make this option more visible.

[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/python/setup.py:353]: SGTM

File: sdks/python/apache_beam/ml/inference/tensorrt_inference.py (hunk 3 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/ml/inference/tensorrt_inference.py:126]: sure, opened: https://github.com/apache/beam/issues/25387

**Pull request merged**